File: tests/firefox-native-context.js
// Real-browser tests; only loaded by firefox-native-context.py's staging addon.
// Copyright (C) 2026 Ignat Remizov
// SPDX-License-Identifier: AGPL-3.0-or-later

"use strict";
const api = browser;
const results = [];
let trace;
const origin = new URL(location.href).searchParams.get('origin');
if (! /^http:\/\/127\.0\.0\.1:\d+$/.test(origin || '')) throw new Error('Loopback fixture origin required');
const sleep = ms => new Promise(resolve => setTimeout(resolve, ms));
function assert(value, message) { if (! value) throw new Error(message); }
async function waitFor(check, label, timeout = 12000) {
  const deadline = Date.now() + timeout;
  let value;
  while (Date.now() < deadline) {
    value = await check();
    if (value) return value;
    await sleep(50);
  }
  throw new Error(`Timed out: ${label}`);
}
async function test(name, run) {
  trace?.push({ time: Date.now(), test: name });
  try { await run(); results.push({ name, pass: true }); }
  catch (err) { results.push({ name, pass: false, error: `${err?.message || err}\n${err?.stack || ''}` }); }
  document.getElementById('results').textContent = JSON.stringify(results, null, 2);
}

try {
  const background = await api.runtime.getBackgroundPage();
  trace = background.__nativeContextTrace;
  const bkgd = await waitFor(() => background.__nativeContextTest, 'background test handle');
  await Promise.race([bkgd.treeLoaded, sleep(15000).then(() => { throw new Error('Background did not initialize'); })]);
  const tree = bkgd.tree;
  if (new URL(location.href).searchParams.get('phase') === 'reload') {
    const { nativeContextCheckpoint: saved } = await api.storage.local.get('nativeContextCheckpoint');
    assert(saved, 'Missing pre-reload checkpoint');
    await test('extension reload retains container scope, group notes, and member identities', async () => {
      assert(bkgd.containers.profileId === saved.profileId, 'Container profile scope changed');
      for (const before of saved.nodes) {
        const node = tree.nodes[before.id];
        assert(node, `Saved node lost: ${before.id}`);
        assert(node.parent.id === before.parent, `Ancestry changed for ${before.id}`);
        if (before.cookieStoreId) {
          assert(node.cookieStoreId === before.cookieStoreId, 'Container reference lost');
          if (before.loaded) assert(node.loaded && node.tabId === before.tabId, 'Live tab binding changed');
        }
        if (before.nativeGroup) {
          assert(node.nativeGroup && node.label === before.label && node.groupColor === before.groupColor,
            'Persistent native group metadata changed');
        }
      }
      assert(tree.root.findNodes(node => node.nativeGroup).length === saved.groupCount, 'Group notes duplicated');
    });
    await test('extension reload keeps synthetic authentication stores separate', async () => {
      for (const [storeId, value] of saved.cookies) {
        const cookie = await api.cookies.get({ url: origin, name: 'test_session', storeId });
        assert(cookie?.value === value, 'Container cookie state changed across reload');
      }
    });
    await test('extension reload reports no unexpected background errors', () => {
      assert(background.__nativeContextErrors.length === 0, background.__nativeContextErrors.join('\n'));
    });
  } else {
  tree.reorderTabsOnCreate = false;
  await api.storage.local.set({ hideCollapsedTabs: false });
  const personal = await api.contextualIdentities.create({ name: 'Personal fixture', color: 'blue', icon: 'fingerprint' });
  const work = await api.contextualIdentities.create({ name: 'Work fixture', color: 'green', icon: 'briefcase' });
  const office = await api.contextualIdentities.create({ name: 'Office fixture', color: 'toolbar', icon: 'briefcase' });
  await bkgd.runSerializedBrowserMutation(() => bkgd.containers.refresh());
  const nativeWindow = await api.windows.create({ url: origin + '/outside', focused: false });
  const win = await waitFor(() => tree.root.getWindowId(nativeWindow.id), 'fixture window');
  const outsideTab = nativeWindow.tabs[0];
  const outside = await waitFor(() => tree.getNodeByTabId(outsideTab.id), 'outside tab');
  const create = async (store, path = '/account') => {
    const tab = await api.tabs.create({ windowId: win.windowId, cookieStoreId: store, url: origin + path, active: false });
    const node = await waitFor(() => tree.getNodeByTabId(tab.id), 'tab attachment');
    await sleep(120);
    return { tab, node };
  };
  const p = await create(personal.cookieStoreId);
  const w = await create(work.cookieStoreId);
  await test('identical URLs retain distinct container IDs and synthetic login cookies', async () => {
    assert(p.node !== w.node, 'Nodes must not be coalesced by URL');
    assert(p.node.cookieStoreId === personal.cookieStoreId, 'Personal identity not captured');
    assert(w.node.cookieStoreId === work.cookieStoreId, 'Work identity not captured');
    assert(p.node.containerName === personal.name && w.node.containerName === work.name, 'Container labels missing');
    await api.cookies.set({ url: origin, name: 'test_session', value: 'personal-only', storeId: personal.cookieStoreId });
    await api.cookies.set({ url: origin, name: 'test_session', value: 'work-only', storeId: work.cookieStoreId });
    assert((await api.cookies.get({ url: origin, name: 'test_session', storeId: personal.cookieStoreId })).value === 'personal-only', 'Personal cookie crossed stores');
    assert((await api.cookies.get({ url: origin, name: 'test_session', storeId: work.cookieStoreId })).value === 'work-only', 'Work cookie crossed stores');
  });

  await outside.moveTo(p.node, p.node.nodes.length, { reason: 'test', skipTabReorder: true, emit: false });
  await w.node.moveTo(outside, outside.nodes.length, { reason: 'test', skipTabReorder: true, emit: false });
  const savedOutside = await p.node.addChild(p.node.nodes.length, { url: origin + '/saved-outside', label: 'Unrelated saved tab' }, { reason: 'test' });
  const nativeGroup = await api.tabs.group({ tabIds: [p.tab.id, w.tab.id], createProperties: { windowId: win.windowId } });
  await api.tabGroups.update(nativeGroup, { title: 'Payments research', color: 'blue' });
  let group;
  await test('native grouping preserves nested members and promotes unrelated descendants', async () => {
    group = await waitFor(() => p.node.getNativeGroupNode(), 'group note');
    await waitFor(() => group.label === 'Payments research' && outside.parent === win
      && w.node.parent === p.node && savedOutside.parent === win, 'settled grouping');
    assert(group.nativeGroup && group.groupId === nativeGroup, 'Wrong native note binding');
    assert(p.node.parent === group && w.node.parent === p.node, 'Member hierarchy flattened');
    assert(savedOutside.parent === win, 'Unrelated saved descendant pulled into group');
  });

  await test('native group rename, color, collapse, and literal Pinned heading synchronize', async () => {
    await api.tabGroups.update(nativeGroup, { title: 'Pinned', color: 'purple', collapsed: true });
    await waitFor(() => group?.label === 'Pinned' && group.groupCollapsed, 'group update');
    assert(group.groupColor === 'purple' && ! group.expanded, 'Color/collapse not reflected');
    assert(! group.isPinnedBranch() && ! p.node.isPinned(), 'Group name interpreted as pin command');
    await group.setExpanded(true, { reason: 'userAction' });
    await waitFor(async () => ! (await api.tabGroups.get(nativeGroup)).collapsed, 'outline expand');
    await group.setNotes('Payments research', group.note, { reason: 'userAction' });
    await waitFor(async () => (await api.tabGroups.get(nativeGroup)).title === 'Payments research', 'outline rename');
  });

  await test('container rename/color/icon changes update the tracked tab', async () => {
    await api.contextualIdentities.update(work.cookieStoreId, { name: 'iremizov fixture', color: 'green', icon: 'briefcase' });
    await waitFor(() => w.node.containerName === 'iremizov fixture', 'identity rename');
    assert(w.node.containerIcon === 'briefcase' && w.node.containerColor === 'green', 'Identity presentation stale');
  });

  const second = await create('firefox-default', '/second-group');
  const secondId = await api.tabs.group({ tabIds: [second.tab.id], createProperties: { windowId: win.windowId } });
  await api.tabGroups.update(secondId, { title: 'Another group', color: 'orange' });
  const secondNote = await waitFor(() => second.node.getNativeGroupNode(), 'second group');
  await test('outline reorder preserves group identities, member nesting, and outsider membership', async () => {
    await secondNote.moveTo(win, 0, { reason: 'userAction' });
    await sleep(300);
    assert((await api.tabs.get(p.node.tabId)).groupId === nativeGroup, 'First native group was replaced');
    assert((await api.tabs.get(w.node.tabId)).groupId === nativeGroup, 'Nested member lost native group');
    assert((await api.tabs.get(second.node.tabId)).groupId === secondId, 'Second group was replaced');
    assert((await api.tabs.get(outside.tabId)).groupId === -1, 'Outside tab accidentally grouped');
    assert(w.node.parent === p.node, 'Group reorder flattened nested members');
  });

  await test('native cross-window group move preserves its note, nesting, and containers', async () => {
    const otherWindow = await api.windows.create({ url: origin + '/other-window', focused: false });
    const other = await waitFor(() => tree.root.getWindowId(otherWindow.id), 'second window attachment');
    try {
      await api.tabGroups.move(nativeGroup, { windowId: otherWindow.id, index: -1 });
      await waitFor(() => group.parent === other && p.node.windowId === otherWindow.id
        && w.node.windowId === otherWindow.id, 'cross-window group synchronization');
      assert(w.node.parent === p.node && p.node.parent === group, 'Native group move flattened members');
      assert(group.groupId === nativeGroup, 'Native move replaced group binding');
      assert((await api.tabs.get(p.node.tabId)).cookieStoreId === personal.cookieStoreId, 'Native move crossed container');
      assert(outside.parent === win && secondNote.parent === win, 'Unrelated tabs followed group');
    } finally {
      await api.tabGroups.move(nativeGroup, { windowId: win.windowId, index: -1 });
      await waitFor(() => group.parent === win && w.node.windowId === win.windowId, 'return native group');
      await api.windows.remove(otherWindow.id);
    }
  });

  await test('outline cross-window group move preserves the native group and its member tree', async () => {
    const otherWindow = await api.windows.create({ url: origin + '/outline-destination', focused: false });
    const other = await waitFor(() => tree.root.getWindowId(otherWindow.id), 'outline destination');
    let failure;
    const captureFailure = async (phase, err) => {
      trace?.push({ time: Date.now(), phase, error: String(err),
        group: group.toDict(), personal: p.node.toDict(), work: w.node.toDict(),
        groups: await api.tabGroups.query({}),
        tabs: (await api.tabs.query({})).map(tab => ({ id: tab.id,
          windowId: tab.windowId, groupId: tab.groupId, index: tab.index })) });
    };
    try {
      await group.moveTo(other, other.nodes.length, { reason: 'userAction' });
      await waitFor(async () => (await api.tabGroups.get(nativeGroup)).windowId === otherWindow.id,
        'outline group moved in browser');
      assert(group.parent === other && p.node.parent === group && w.node.parent === p.node, 'Outline move flattened tree');
      assert((await api.tabs.get(w.node.tabId)).cookieStoreId === work.cookieStoreId, 'Outline move crossed container');
      assert((await api.tabs.get(outside.tabId)).windowId === win.windowId, 'Outside tab moved');
    } catch (err) {
      failure = err;
      await captureFailure('outbound outline move', err);
    } finally {
      try {
        await group.moveTo(win, win.nodes.length, { reason: 'userAction' });
        await waitFor(async () => (await api.tabGroups.get(nativeGroup)).windowId === win.windowId,
          'return outline group');
        await api.windows.remove(otherWindow.id);
      } catch (err) {
        await captureFailure('return outline move', err);
        failure ||= err;
      }
    }
    if (failure) throw failure;
  });

  await test('saved grouped tab resumes in its exact container and existing native group', async () => {
    const originalId = p.node.tabId;
    await p.node.unload({ reason: 'userAction' });
    await waitFor(() => ! p.node.loaded && ! p.node.tabId, 'unload container tab');
    const response = await bkgd.bkgd_loadSavedNode({ nodeId: p.node.id, discarded: false });
    assert(! response.error && response.result === 'ok', JSON.stringify(response));
    await waitFor(() => p.node.loaded && p.node.tabId !== originalId && p.node.groupId === nativeGroup, 'restore grouped container tab');
    const tab = await api.tabs.get(p.node.tabId);
    assert(tab.cookieStoreId === personal.cookieStoreId && tab.groupId === nativeGroup, 'Restore crossed identity/group');
    assert((await api.cookies.get({ url: origin, name: 'test_session', storeId: tab.cookieStoreId })).value === 'personal-only', 'Restored authentication store changed');
  });

  await test('deleted container restore stays saved and creates no replacement tab', async () => {
    const o = await create(office.cookieStoreId, '/deleted-container');
    await o.node.setNotes('Missing container example', undefined, { reason: 'test' });
    await o.node.unload({ reason: 'userAction' });
    await waitFor(() => ! o.node.loaded, 'unload office');
    await api.contextualIdentities.remove(office.cookieStoreId);
    await waitFor(() => o.node.containerMissing, 'deleted identity metadata');
    const before = (await api.tabs.query({})).length;
    const response = await bkgd.bkgd_loadSavedNode({ nodeId: o.node.id });
    assert(response.error && o.node.restoreError, 'Missing identity failure not visible');
    assert((await api.tabs.query({})).length === before && ! o.node.loaded, 'Opened fallback tab');
  });

  await test('initial tab of a new saved window opens in the saved container and group', async () => {
    const savedWindow = await tree.root.addChild(tree.root.nodes.length, { type: 'window', label: 'Restored container window', incognito: false }, { reason: 'test' });
    const savedGroup = await savedWindow.addChild(0, { nativeGroup: true, label: 'Saved group', groupColor: 'green', groupCollapsed: true }, { reason: 'test' });
    const node = await savedGroup.addChild(0, { url: origin + '/new-window', cookieStoreId: personal.cookieStoreId,
      containerProfileId: bkgd.containers.profileId, containerName: personal.name, containerColor: personal.color, containerIcon: personal.icon }, { reason: 'test' });
    const response = await bkgd.bkgd_loadSavedNode({ nodeId: node.id, discarded: false });
    assert(! response.error && response.result === 'ok', JSON.stringify(response));
    try {
      await waitFor(() => node.loaded && node.tabId && savedGroup.groupId !== undefined, 'new-window container restore');
    } catch (err) {
      throw new Error(`${err.message}\n${JSON.stringify({
        node: node.toDict(), group: savedGroup.toDict(), window: savedWindow.toDict(),
        pending: bkgd.nodesLoading.map(item => item.id),
        browsers: await api.windows.getAll({ populate: true }),
        candidates: Object.values(tree.nodes).filter(item => item.url?.endsWith('/new-window')).map(item => item.toDict())
      }, null, 2)}`);
    }
    assert((await api.tabs.get(node.tabId)).cookieStoreId === personal.cookieStoreId, 'Initial URL loaded in default store');
    assert((await api.tabGroups.get(savedGroup.groupId)).title === 'Saved group', 'Group was not recreated');
    await api.windows.remove(node.windowId);
    await waitFor(() => ! node.loaded, 'close restored window');
  });

  await test('native ungroup promotes a member without moving its remaining grouped descendants', async () => {
    await api.tabs.ungroup(p.node.tabId);
    // Reconciliation persists/broadcasts several moves. Observe the complete
    // invariant, not the instant between promoting the parent and its child.
    await waitFor(() => ! p.node.getNativeGroupNode()
      && w.node.getNativeGroupNode() === group, 'native ungroup');
    assert(w.node.getNativeGroupNode() === group, 'Still-grouped child followed its ungrouped parent');
    await api.tabs.group({ groupId: nativeGroup, tabIds: [p.node.tabId] });
    await waitFor(() => p.node.getNativeGroupNode() === group, 'regroup');
  });

  await test('closing and restoring a native window preserves containers and group note identities', async () => {
    const windowNodeId = win.id, groupNodeId = group.id, personalNodeId = p.node.id, workNodeId = w.node.id;
    const oldWindowId = win.windowId;
    await api.windows.remove(oldWindowId);
    await waitFor(() => ! win.loaded && ! p.node.loaded && ! w.node.loaded, 'closed native window');
    const closed = await waitFor(async () => (await api.sessions.getRecentlyClosed())
      .find(item => item.window?.tabs?.some(tab => tab.url === origin + '/account')), 'closed session');
    const restored = await api.sessions.restore(closed.window.sessionId);
    await waitFor(() => tree.nodes[personalNodeId]?.loaded && tree.nodes[workNodeId]?.loaded, 'native session tabs restored');
    await sleep(700);
    assert(tree.nodes[windowNodeId] === win && win.windowId === restored.window.id, 'Window history duplicated');
    assert(tree.nodes[groupNodeId] === group && w.node.getNativeGroupNode() === group && p.node.getNativeGroupNode() === group, 'Group note replaced/duplicated on session restore');
    assert((await api.tabs.get(p.node.tabId)).cookieStoreId === personal.cookieStoreId, 'Personal store lost on native restore');
    assert((await api.tabs.get(w.node.tabId)).cookieStoreId === work.cookieStoreId, 'Work store lost on native restore');
  });

  await test('background has no unexpected errors during native context lifecycle', async () => {
    const errors = background.__nativeContextErrors || [];
    assert(errors.length === 0, errors.join('\n'));
  });
  // The driver reloads the test addon, then checks this against IndexedDB's
  // reloaded tree and the browser's original live tabs, not in-memory objects.
  await bkgd.runSerializedBrowserMutation(() => bkgd.tabGroups.sync());
  await api.storage.local.set({ nativeContextCheckpoint: {
    profileId: bkgd.containers.profileId, focusNodeId: win.id,
    groupCount: tree.root.findNodes(node => node.nativeGroup).length,
    nodes: [group, p.node, w.node, secondNote, savedOutside].map(node => node.toDict()),
    cookies: [[personal.cookieStoreId, 'personal-only'], [work.cookieStoreId, 'work-only']]
  } });
  }
} catch (err) {
  results.push({ name: 'integration setup', pass: false, error: `${err?.message || err}\n${err?.stack || ''}` });
} finally {
  window.__TEST_RESULTS__ = { results, passCount: results.filter(item => item.pass).length,
    failCount: results.filter(item => ! item.pass).length };
  if (window.__TEST_RESULTS__.failCount) window.__TEST_RESULTS__.trace = trace;
  document.getElementById('results').textContent = JSON.stringify(window.__TEST_RESULTS__, null, 2);
  window.__TEST_DONE__ = true;
}

File: tests/firefox-native-context.py
#!/usr/bin/env python3
"""Exercise the real extension in a disposable Firefox/Geckodriver profile.

No Selenium dependency. Only synthetic loopback pages/cookies are used. The
test-only background handle and host permission are inserted in a staging
copy, never in the shipped extension. No existing Firefox profile is opened.
"""
# Copyright (C) 2026 Ignat Remizov
# SPDX-License-Identifier: AGPL-3.0-or-later

import argparse
import base64
import http.server
import json
import os
from pathlib import Path
import shutil
import socket
import subprocess
import tempfile
import threading
import time
import urllib.error
import urllib.parse
import urllib.request
import zipfile

ROOT = Path(__file__).resolve().parents[1]
ADDON_ID = "native-context-test@tktsto.invalid"


class FixtureServer(http.server.BaseHTTPRequestHandler):
    def do_GET(self):
        if self.path.split("?", 1)[0] == "/favicon.ico":
            self.send_response(204)
            self.end_headers()
            return
        data = b"<!doctype html><title>Synthetic account fixture</title><p>Local container test.</p>"
        self.send_response(200)
        self.send_header("Content-Type", "text/html; charset=utf-8")
        self.send_header("Cache-Control", "no-store")
        self.send_header("Content-Length", str(len(data)))
        self.end_headers()
        self.wfile.write(data)

    def log_message(self, *_args):
        pass


def free_port():
    with socket.socket() as listener:
        listener.bind(("127.0.0.1", 0))
        return listener.getsockname()[1]


def stage_extension(work, trace_api=False, expose_views=False):
    staged = work / "extension"
    staged.mkdir()
    for source in ROOT.glob("*.js"):
        shutil.copy2(source, staged / source.name)
    for name in ("_locales", "bkgd", "common", "docs", "img", "options", "themes", "view"):
        shutil.copytree(ROOT / name, staged / name)
    manifest = json.loads((ROOT / "manifest-ff.json").read_text())
    manifest["browser_specific_settings"]["gecko"]["id"] = ADDON_ID
    manifest["host_permissions"] = ["http://127.0.0.1/*"]
    (staged / "manifest.json").write_text(json.dumps(manifest, indent=2))
    source_path = staged / "bkgd/bkgd.js"
    source = source_path.read_text()
    marker = "  const bkgd = new Bkgd();\n  bkgd.init();"
    if source.count(marker) != 1:
        raise RuntimeError("Background test-hook location changed")
    source = source.replace(marker, """  const bkgd = new Bkgd();
  globalThis.__nativeContextTest = bkgd;
  globalThis.__TKTSTO_TEST_QUIET__ = true;
  globalThis.__nativeContextErrors = [];
  globalThis.__nativeContextTrace = [];
  if (__TRACE_NATIVE_CONTEXT__) {
  for (const [namespace, methods] of [['tabs', ['move', 'group', 'ungroup']], ['tabGroups', ['move', 'update']]]) {
    for (const method of methods) {
      const original = api[namespace][method];
      api[namespace][method] = (...args) => {
        const entry = { time: Date.now(), call: namespace + '.' + method, args,
          stack: new Error().stack };
        globalThis.__nativeContextTrace.push(entry);
        return original(...args).then(value => { entry.result = value; return value; },
          error => { entry.error = String(error); throw error; });
      };
    }
  }
  for (const method of ['onTabAttached', 'onTabMoved']) {
    const original = bkgd[method].bind(bkgd);
    bkgd[method] = async (...args) => {
      const entry = { time: Date.now(), event: method, args };
      globalThis.__nativeContextTrace.push(entry);
      try { return await original(...args); }
      finally {
        const node = bkgd.tree?.getNodeByTabId(args[0]);
        entry.node = node ? { id: node.id, parent: node.parent?.id,
          windowId: node.windowId, groupId: node.groupId } : null;
      }
    };
  }
  }
  const originalError = console.error.bind(console);
  console.error = (...args) => {
    globalThis.__nativeContextErrors.push(args.map(String).join(' '));
    originalError(...args);
  };
  bkgd.init();""")
    source_path.write_text(source.replace('__TRACE_NATIVE_CONTEXT__', json.dumps(trace_api)))
    (staged / "tests").mkdir()
    for name in ("firefox-native-context.html", "firefox-native-context.js"):
        shutil.copy2(ROOT / "tests" / name, staged / "tests" / name)
    if expose_views:
        view_path = staged / "view/view.js"
        view_source = view_path.read_text()
        marker = "  const tree = new TreeView();"
        if view_source.count(marker) != 1:
            raise RuntimeError("View test-hook location changed")
        view_path.write_text(view_source.replace(marker,
            marker + "\n  window.__isolatedTestTree = tree;"))
    package = work / "test-extension.zip"
    with zipfile.ZipFile(package, "w", zipfile.ZIP_DEFLATED) as archive:
        for file in staged.rglob("*"):
            if file.is_file():
                archive.write(file, file.relative_to(staged))
    return package


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--geckodriver", default=shutil.which("geckodriver"))
    parser.add_argument("--output", type=Path)
    parser.add_argument("--trace-api", action="store_true", help="Trace synthetic group API calls when diagnosing a failure")
    args = parser.parse_args()
    if not args.geckodriver:
        parser.error("geckodriver is required")
    # Ubuntu Snap Firefox and its driver share this path; neither necessarily
    # shares the host's /tmp. It is always a NEW directory, not a user profile.
    parent = Path.home() / "snap/firefox/common" if "/snap/" in args.geckodriver else Path.home()
    parent.mkdir(parents=True, exist_ok=True)
    work = Path(tempfile.mkdtemp(prefix="tktsto-context-test-", dir=parent))
    output = args.output or work
    output.mkdir(parents=True, exist_ok=True)
    print(f"Disposable test data: {work}", flush=True)
    package = stage_extension(work, trace_api=args.trace_api)
    server = http.server.ThreadingHTTPServer(("127.0.0.1", 0), FixtureServer)
    threading.Thread(target=server.serve_forever, daemon=True).start()
    origin = f"http://127.0.0.1:{server.server_port}"
    port = free_port()
    base = f"http://127.0.0.1:{port}"
    log = (output / "firefox-driver.log").open("w")
    driver = subprocess.Popen([args.geckodriver, "--port", str(port), "--host", "127.0.0.1",
                               "--profile-root", str(work), "--allow-system-access"],
                              stdout=log, stderr=subprocess.STDOUT)
    session = None

    def request(method, endpoint, data=None, timeout=100):
        body = json.dumps(data).encode() if data is not None else None
        req = urllib.request.Request(base + endpoint, data=body, method=method,
                                     headers={"Content-Type": "application/json"})
        try:
            with urllib.request.urlopen(req, timeout=timeout) as response:
                value = json.load(response).get("value")
        except urllib.error.HTTPError as exc:
            raise RuntimeError(exc.read().decode()) from exc
        if isinstance(value, dict) and value.get("error"):
            raise RuntimeError(json.dumps(value))
        return value

    def command(method, endpoint, data=None):
        return request(method, f"/session/{session}/{endpoint}", data)

    try:
        deadline = time.monotonic() + 15
        while True:
            if driver.poll() is not None:
                raise RuntimeError("geckodriver exited; see firefox-driver.log")
            try:
                request("GET", "/status", timeout=1)
                break
            except (OSError, RuntimeError):
                if time.monotonic() >= deadline:
                    raise RuntimeError("geckodriver did not start")
                time.sleep(0.1)
        capabilities = request("POST", "/session", {"capabilities": {"alwaysMatch": {
            "browserName": "firefox", "moz:firefoxOptions": {
                **({"binary": os.environ["FIREFOX_BIN"]} if os.environ.get("FIREFOX_BIN") else {}),
                "args": ["-headless", "-no-remote"],
                "prefs": {"privacy.userContext.enabled": True, "browser.tabs.groups.enabled": True,
                          "browser.startup.page": 0, "datareporting.policy.dataSubmissionEnabled": False}
            }
        }}})
        session = capabilities["sessionId"]
        print(f"Firefox {capabilities['capabilities']['browserVersion']}; isolated WebDriver session", flush=True)
        command("POST", "timeouts", {"script": 90000, "pageLoad": 30000})
        command("POST", "moz/addon/install", {"path": str(package), "temporary": True})
        command("POST", "moz/context", {"context": "chrome"})
        extension_url = command("POST", "execute/sync", {
            "script": "return WebExtensionPolicy.getByID(arguments[0]).getURL('');", "args": [ADDON_ID]})
        command("POST", "moz/context", {"context": "content"})
        test_url = extension_url + "tests/firefox-native-context.html?" + urllib.parse.urlencode({"origin": origin})
        command("POST", "url", {"url": test_url})
        wait_result = {"script": """
          const done = arguments[arguments.length - 1];
          const timer = setInterval(() => {
            if (window.__TEST_DONE__) {
              clearInterval(timer);
              done(window.__TEST_RESULTS__);
            }
          }, 100);
        """, "args": []}
        result = command("POST", "execute/async", wait_result)
        (output / "firefox-native-context-results.json").write_text(json.dumps(result, indent=2))
        print(f"Lifecycle: {result['passCount']} passed, {result['failCount']} failed", flush=True)
        if not result["failCount"]:
            # Reload only our temporary addon in this disposable profile.
            # Navigation waits for its new background initialization below.
            command("POST", "execute/sync", {"script": "setTimeout(() => browser.runtime.reload(), 0); return true;", "args": []})
            time.sleep(0.8)
            # Firefox closes the old addon page during reload. Select one of
            # this disposable session's remaining tabs, then create a fresh
            # harness window rather than reusing the discarded context.
            handles = command("GET", "window/handles")
            command("POST", "window", {"handle": handles[0]})
            harness = command("POST", "window/new", {"type": "window"})
            command("POST", "window", {"handle": harness["handle"]})
            command("POST", "url", {"url": test_url + "&phase=reload"})
            reloaded = command("POST", "execute/async", wait_result)
            result["results"].extend(reloaded["results"])
            result["passCount"] += reloaded["passCount"]
            result["failCount"] += reloaded["failCount"]
        (output / "firefox-native-context-results.json").write_text(json.dumps(result, indent=2))
        for item in result.get("results", []):
            print(f"{'PASS' if item['pass'] else 'FAIL'}: {item['name']} {item.get('error', '')}", flush=True)
        print(json.dumps({key: result[key] for key in ("passCount", "failCount")}), flush=True)
        # Render the actual production TreeView, not a mock, with synthetic
        # tabs and containers from this test profile.
        command("POST", "window/rect", {"width": 760, "height": 1000})
        command("POST", "url", {"url": extension_url + "view/sidepanel.html"})
        time.sleep(1.5)
        # Scroll the real tree to the fixture window; don't hide or restyle any
        # production DOM. The tutorial in the first window is otherwise taller
        # than the viewport and obscures the container/group rows.
        visible = command("POST", "execute/async", {"script": """
          const done = arguments[arguments.length - 1];
          browser.storage.local.get('nativeContextCheckpoint').then(({nativeContextCheckpoint: saved}) => {
            const id = 'node' + saved?.focusNodeId;
            if (!document.getElementById(id)) document.getElementById('view-scope-btn')?.click();
            const deadline = Date.now() + 5000;
            const timer = setInterval(() => {
              const target = document.getElementById(id);
              if (target) {
                clearInterval(timer);
                target.scrollIntoView({block: 'start'});
                done(true);
              } else if (Date.now() > deadline) {
                clearInterval(timer); done(false);
              }
            }, 50);
          }, error => done(String(error)));
        """, "args": []})
        screenshot = command("GET", "screenshot")
        (output / "firefox-native-context.png").write_bytes(base64.b64decode(screenshot))
        if result["failCount"]:
            raise RuntimeError("Firefox native-context integration tests failed")
        if visible is not True:
            raise RuntimeError("Production sidebar could not display the synthetic fixture window")
    finally:
        if session:
            try:
                request("DELETE", f"/session/{session}", timeout=15)
            except (OSError, RuntimeError):
                pass
        try:
            driver.terminate()
        except PermissionError:
            # Ubuntu's Snap signal policy can reject the named Codex AppArmor
            # peer even for its own child. Ask the same user's service manager
            # to terminate precisely this test-owned PID, without sudo.
            subprocess.run(["systemd-run", "--user", "--wait", "--collect", "--quiet",
                            "/usr/bin/kill", "-TERM", str(driver.pid)], check=True)
        try:
            driver.wait(timeout=5)
        except subprocess.TimeoutExpired:
            driver.kill()
            driver.wait(timeout=5)
        server.shutdown()
        server.server_close()
        log.close()
        # Leave the disposable staging directory and diagnostics available for
        # inspection. No process uses it after this function exits.
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
